src/engine/renderer/glsl_source/lightMapping_fp.glsl

```diff
@@ -27,6 +27,7 @@ uniform sampler2D	u_MaterialMap;
 uniform sampler2D	u_GlowMap;
 
 uniform float		u_AlphaThreshold;
+uniform float u_LightFactor;
```

Comment:
Maybe you could optimize it a bit by inverting (x -> 1/x) this uniform so that you can use multiplication instead of division. Also by setting it to 0 instead of 1 when there is no need for multiplications so then the operations can be skipped.

Reply:
Nice catch. I now do the division once at map load time instead, and replaced all the GLSL divisions by a multiplication.

It is assumed most GPUs do not know how to do jumps and then never skip a computation in a test but just discard the result if the test is false, so `something *= a` with `a = 1` is better than `if ( a != 0 ) { something *= a; }` with `a = 0` for when there is nothing to multiply, as we avoid the test instruction in all cases and we cannot avoid the multiplication in all cases.

Reply:
> It is assumed most GPUs do not know how to do jumps and then never skip a computation in a test but just discard the result if the test is false

My concept is a bit different. I think they can jump, but all threads must execute the same code. With `if (x) { foo(); } else { bar(); }` if `x` is true at least once and false at least once, all threads have to execute both branches. But if `x` is false in all threads it can jump directly to `bar()`.

I agree that `x *= 1;` may very well be faster than `if ( false ) { x *= 1; }`. But my contention is that `if ( false ) { x *= 1; }` is likely faster than `if ( true ) { x *= 1; }`. (Here by `false`, `true` and `1` I mean expressions derived from uniforms that always have the same value, rather than compile-time constants.)

Reply:
Hmm, maybe, but, that complexifies the code for the reader for a benefit that may be very minor. Maybe @VReaperV has an opinion on it?

Reply:
> Here by false, true and 1 I mean expressions derived from uniforms that always have the same value

If it's based only on uniforms and constant expressions (which it is unless I'm missing something), then it will be statically uniform so there wouldn't be any divergence and the unused code path will likely be eliminated.

Also, keep in mind that a compiler might optimise something like `if( condition > 0.0f ) { value *= condition }` into `value *= condition > 0.0f ? condition : 1` anyway.

At the end of the day though, trying to optimise a single vec4 multiplication that is done once per fragment is a waste of time. (If you really want to know which one will be faster you'd need to profile both and it would be hw/driver specific and you likely won't see any significant difference, and it would just be a giant waste of time)

Reply:
Thank you for the explanations! ❤️